File: cosmoauthorization/src/java/it/csi/cosmo/cosmoauthorization/business/service/TagsService.java
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package it.csi.cosmo.cosmoauthorization.business.service;

import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaTagRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaTagRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.TagResponse;
import it.csi.cosmo.cosmoauthorization.dto.rest.TagsResponse;

public interface TagsService {

  /**
   * Metodo per eliminare un tag dal database
   *
   * @param id e' l'identificativo numerico del tag da eliminare
   */
  void deleteTag(Long id);

  /**
   * Metodo che restituisce una lista di tags validi
   *
   * @param filter
   * @return una lista di tags validi
   */
  TagsResponse getTags(String filter);

  /**
   * Metodo che restituisce un tag specifico
   *
   * @param id del tag che deve essere cercato nel database
   * @return tag presente nel database
   */
  TagResponse getTagById(Long id);

  /**
   * Metodo che crea un tag sul database
   *
   * @param body e' il tag che deve essere creato sul database
   * @return il tag creato sul database
   */
  TagResponse postTag(CreaTagRequest body);

  /**
   * Metodo che aggiorna un tag gia' esistente nel database
   *
   * @param id e' l'id fisico del tag che deve essere aggiornato
   * @param body e' il tag che deve essere aggiornato
   * @return il tag aggiornato nel database
   */
  TagResponse updateTag(Long id, AggiornaTagRequest body);
}

File: cosmowcl/src/app/shared/services/tags.service.ts
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { AggiornaTagRequest } from '../models/api/cosmoauthorization/aggiornaTagRequest';
import { CreaTagRequest } from '../models/api/cosmoauthorization/creaTagRequest';
import { TagResponse } from '../models/api/cosmoauthorization/tagResponse';
import { TagsResponse } from '../models/api/cosmoauthorization/tagsResponse';
import { TipoTag } from '../models/api/cosmoauthorization/tipoTag';
import { ApiUrls } from '../utilities/apiurls';
import { Utils } from '../utilities/utilities';

@Injectable({
  providedIn: 'root'
})
export class TagsService {

  constructor(
    private http: HttpClient
  ) { }

  getTipiTag(): Observable<TipoTag[]> {
    return this.http.get<TipoTag[]>(ApiUrls.GET_TIPI_TAGS);
  }

  /** GET: Restituisce una lista tags */
  getTags(filter: string): Observable<TagsResponse> {
    const options = filter ?
      { params: new HttpParams().set('filter', filter) } : {};
    return this.http.get<TagsResponse>(ApiUrls.TAGS, options);
  }

  get(id: string): Observable<TagResponse> {
    Utils.require(id);
    return this.http.get<TagResponse>(ApiUrls.TAG_ID.replace('{id}', id.toString()));
  }

  deleteTag(id: number): Observable<void> {
    Utils.require(id);
    return this.http.delete<void>(ApiUrls.TAG_ID.replace('{id}', id.toString()));
  }

  createTag(payload: CreaTagRequest): Observable<TagResponse> {
    Utils.require(payload, 'payload');
    return this.http.post<TagResponse>(ApiUrls.TAGS, payload);
  }

  updateTag(id: number, payload: AggiornaTagRequest): Observable<TagResponse> {
    Utils.require(id);
    return this.http.put<TagResponse>(ApiUrls.TAG_ID.replace('{id}', id.toString()), payload);
  }
}



